test/system/src/main/java/io/pravega/test/system/framework/services/docker/DockerBasedService.java

```diff
@@ -0,0 +1,158 @@
+                dockerClient.createNetwork(NetworkConfig.builder().driver("overlay").name("docker-network").build());
+            }
+        } catch (DockerException | InterruptedException e) {
+            log.error("Unable to create network {}", e);
+        }
+        this.serviceName = serviceName;
+        }
+
+
+    @Override
+    public String getID() {
+        Service.Criteria criteria = Service.Criteria.builder().serviceName(this.serviceName).build();
+        String serviceId = null;
+        try {
+            List<Service> serviceList = dockerClient.listServices(criteria);
+            serviceId = serviceList.get(0).id();
+        }  catch (DockerException | InterruptedException  e) {
+            log.error("Unable to get service id {}", e);
+        }
+        return serviceId;
+    }
+
+    @Override
+    public boolean isRunning() {
+        boolean value = false;
+        try {
+            //list the service with filter 'serviceName'
+            Service.Criteria criteria = Service.Criteria.builder().serviceName(this.serviceName).build();
+            List<Service> serviceList = dockerClient.listServices(criteria);
+
+            for (int i = 0; i < serviceList.size(); i++) {
+                String serviceId = serviceList.get(i).id();
+                if (!serviceId.isEmpty()) {
+                    value = true;
+                    break;
+                }
+            }
+        } catch (DockerException | InterruptedException e) {
+            log.error("Unable to list docker services {}", e);
+        }
+        return value;
+    }
+
+    CompletableFuture<Void> waitUntilServiceRunning() {
+        log.debug("IS RUNNING {}", isRunning());
+        return FutureHelpers.loop(() -> !isRunning(), //condition
+                () -> FutureHelpers.delayedFuture(Duration.ofSeconds(5), executorService),
+                executorService);
+    }
+
+    String setSystemProperty(final String propertyName, final String propertyValue) {
+        return new StringBuilder().append(" -D").append(propertyName).append("=").append(propertyValue).toString();
+    }
+
+    @Override
+    public List<URI> getServiceDetails() {
+        Service.Criteria criteria = Service.Criteria.builder().serviceName(this.serviceName).build();
+        List<URI> uriList = new ArrayList<>();
+        try {
+            List<Service> serviceList = dockerClient.listServices(criteria);
+
+            for (int i = 0; i < serviceList.size(); i++) {
+                URI uri = URI.create("tcp://" + dockerClient.getHost() + ":"
+                        + serviceList.get(i).spec().endpointSpec()
+                        .ports().get(i).targetPort());
+                uriList.add(uri);
+            }
+        }  catch (InterruptedException | DockerException e) {
+            log.error("Unable to list service details {}", e);
+        }
+        return uriList;
+    }
+
+    @Override
+    public void scaleService(final int instanceCount, final boolean wait) {
```

Comment:
Rather than a boolean to block or not, it should return a future which the caller can decide to call get() on or not.

Reply:
done.